In [1]:
import pandas as pd
import numpy as np
import os
from matplotlib import pyplot as plt
import numpy as np
import math
from sklearn.preprocessing import StandardScaler

In [17]:
iris = pd.read_csv('../../assets/datasets/iris.csv')

In [19]:
iris.head()

SepalLength  SepalWidth  PetalLength  PetalWidth         Name
0          5.1         3.5          1.4         0.2  Iris-setosa
1          4.9         3.0          1.4         0.2  Iris-setosa
2          4.7         3.2          1.3         0.2  Iris-setosa
3          4.6         3.1          1.5         0.2  Iris-setosa
4          5.0         3.6          1.4         0.2  Iris-setosa

## Step 1: Split the set into two sets

"X" will be the data and "Y" will be the class labels

In [22]:
X = iris.iloc[:,0:4].values
y = iris.iloc[:,4].values

## Step 2: Explore the Data

Next - Let's plot! You can use any plotting library of your choice, but be sure to explore all of the data. 

## Step 3: Dimensionality Reduction

First, standarize the data. While the Iris data attributes are all measured in the same units (cm), this is a worthwhile step for optimization and good practice for more unruly datasets!

In [24]:
X_standard = StandardScaler().fit_transform(X)

Now, let's set up our data for decomposition by creating a covariance matrix

Now, decompose the the covariance matrix

In [26]:
cov_mat = np.cov(X_standard.T)
eigenValues, eigenVectors = np.linalg.eig(cov_mat)

Check the eigenvalues and eigenvectors

In [28]:
eigenValues

array([ 2.93035378,  0.92740362,  0.14834223,  0.02074601])

In [29]:
eigenVectors

array([[ 0.52237162, -0.37231836, -0.72101681,  0.26199559],
       [-0.26335492, -0.92555649,  0.24203288, -0.12413481],
       [ 0.58125401, -0.02109478,  0.14089226, -0.80115427],
       [ 0.56561105, -0.06541577,  0.6338014 ,  0.52354627]])

The eigenvectors with the lowest eigenvalues can be dropped

In [37]:
eig_pairs = [(np.abs(eigenValues[i]), eigenVectors[:,i]) for i in range(len(eigenValues))]
eig_pairs.sort()
eig_pairs.reverse()
for i in eig_pairs:
    print(i)

(2.9303537755893174, array([ 0.52237162, -0.26335492,  0.58125401,  0.56561105]))
(0.92740362151734212, array([-0.37231836, -0.92555649, -0.02109478, -0.06541577]))
(0.14834222648163944, array([-0.72101681,  0.24203288,  0.14089226,  0.6338014 ]))
(0.020746013995595943, array([ 0.26199559, -0.12413481, -0.80115427,  0.52354627]))


Calculate the explained variance

In [34]:
tot = sum(eigenValues)
var_exp = [(i / tot)*100 for i in sorted(eigenValues, reverse=True)]
cum_var_exp = np.cumsum(var_exp)

In [35]:
cum_var_exp

array([  72.77045209,   95.80097536,   99.48480732,  100.        ])

What does the explained variance tell us?

73% of variance is explained by first principal component